### ISSUE TITLE
Feature Request - Interlinked cabs

### ISSUE THREAD

**Issue by Montygar** [p1pkin/demul#399]
I would like to be able to either interlink 2 computers running demul. Many games such as initialD have competitive modes which rely on 1 player per cab.

**Comment by p1pkin** [comment]
that number of NAOMI games supports cabinet linking ?
I highly doubt there "many games", my guess is 10 or so

**Comment by Montygar** [comment]
Well there are quite a few. Certainly enough to merit looking into it even if it is low priority.
From what I can tell the following games support linking cabinets:

Gundam 
OutTrigger 
Initial D 
Heavy Metal 
Mazan 
Ninja Assault 
Biohazard 
Virtual On 
Wave Runner 
F355 Twin (maybe others of this series) 
Club Kart 
Ring Out 4x4 (I think) 
Slashout (I think) 
Spikers Battle (I think) 
Wild Riders (Not sure) 
Spawn
Tetris

**Comment by p1pkin** [comment]
you can't tell, you can copy/paste from Arcade-Otaku :)
above list far from correct, there many games doesn't have link features.

**Comment by p1pkin** [comment]
anyway, communication board emulation is planned, and mostly implemented already.
but currently it works at local PC only, not sure its possible at all to link several machines over internet, because games require very low connection latency and will not work or hang if it is not.

**Comment by Montygar** [comment]
Fair enough. I was going to aim for a multimonitor dual cab anyway to it might be cool to run 2 clients in parallel. I assume the controller plug in would have to be rewritten to allow 2 seperate player 1's

**Comment by vkedwardli** [comment]
Seems the latest DEmul v0.7A 180816 does ship with a partial working communication board emulation?
Was trying with Gundam DX, during the MASTER / SLAVE screen they can detect each other, but stuck at a black screen instead of the "NETWORK OK. ID=X" screen.

**Comment by cooperate** [comment]
This topic is related to my post about Derby Owners Club.  Ideally you would want to emulate the separate instances on different machines.  For games like Derby Owners Club though it seems that only one client can connect at max to the master screen.  Derby Owners Club should allow up to 8 clients to connect at once to the master screen.

**Comment by vkedwardli** [comment]
@cooperate For #421, did you connect 1 master and 1 slave successfully? Thanks

**Comment by cooperate** [comment]
@vkedwardli 
Yes I did.  

The process is this:
Launch main screen instance of Derby Owners Club.  Let's use the world edition EX in this example.

1)So launch Demul->File->Run Naomi->derbyocwm (Main)(Rev. D)

Demul should launch two windows for both "master" screens in the game.  Wait until the main screen says "Checking Network"

2)Launch another instance of Demul->File-Run Naomi->derbyocw

This will launch the client or "satellite" instance of Derby Owners Club.  After the bios it will show the blue screen and connect with this message.

https://i.imgur.com/4WVlej6.png

Followed by this message:

https://i.imgur.com/ZWQkFiM.png

Afterwards you should be able to play the game.  

https://i.imgur.com/0d0WmNE.png

Next up getting more than one "slave" or satellite instance to connect to the master.  So close I feel!

**Comment by vkedwardli** [comment]
@cooperate thank you for the information, are you using version v0.7A 180816?

**Comment by cooperate** [comment]
@vkedwardli no v0.7 171011

**Comment by vkedwardli** [comment]
@cooperate oops didn't know there is a new WIP release, thanks

![1](https://i.imgur.com/tMiw6rJ.png)
![2](https://i.imgur.com/WuX1BTP.png)
![3](https://i.imgur.com/evFRdMc.png)
![4](https://i.imgur.com/n2ljysg.png)

I'm running 2 demul from a different folder, tried with Naomi `epr-21576g`/`epr-21576h`; DIMM 3.17 / 4.01 ; DIMM / Net-DIMM / Net-only DIMM

@p1pkin FYI, this is my test results for 171011 WIP, for the previous 180816 I always get a black screen, now I get different result randomly (including both black screens also)

Thanks for the hard work 👍

**Comment by vkedwardli** [comment]
Not sure if anything related to communication board is updated in `180428`, I'm getting correct Network ID assignment while testing with 1 master and 1 slave, though still cannot load the game after seeing `ID=1` and `ID=2`

**Comment by vkedwardli** [comment]
bump 😁

**Comment by MortyLau** [comment]
> 
> 
> @cooperate oops didn't know there is a new WIP release, thanks
> 
> ![1](https://camo.githubusercontent.com/7dcd526f1e582b85c94b7b257cc35bd22edc1e92/68747470733a2f2f692e696d6775722e636f6d2f744d697736724a2e706e67)
> ![2](https://camo.githubusercontent.com/8dd93cd3b355e16b450a5cfa483214861422bf9a/68747470733a2f2f692e696d6775722e636f6d2f577558314254502e706e67)
> ![3](https://camo.githubusercontent.com/e2f421c0eafcb43854815dcca87ec2303d2111dd/68747470733a2f2f692e696d6775722e636f6d2f65764652644d632e706e67)
> ![4](https://camo.githubusercontent.com/04f9f5bcc84d30abfbd3988ebfb2d92d5adba322/68747470733a2f2f692e696d6775722e636f6d2f6e326c6a7973672e706e67)
> 
> I'm running 2 demul from a different folder, tried with Naomi `epr-21576g`/`epr-21576h`; DIMM 3.17 / 4.01 ; DIMM / Net-DIMM / Net-only DIMM
> 
> @p1pkin FYI, this is my test results for 171011 WIP, for the previous 180816 I always get a black screen, now I get different result randomly (including both black screens also)
> 
> Thanks for the hard work 👍

how can you do this? sorry, i cant find a way to the screen "checking network". Thanks

**Comment by vkedwardli** [comment]
@MortyLau you can set the master/slave mode in the game service menu

**Comment by MortyLau** [comment]
> 
> 
> @MortyLau you can set the master/slave mode in the game service menu

Thank you for your reply. But i dont know where it is, i only know how to enable the dimm and choose net-dimm. Would you please teach how to use this emulator?

**Comment by vkedwardli** [comment]
@MortyLau Actually as I've mentioned above: the communication board emulation does not work for Gundam DX.

So even if you can found the option to set master/slave, it won't work.
Meanwhile I've focusing on rebuilding the game server for the DC/PS2 version.
If you are good at reverse engineering feel free to participate https://github.com/inada-s/gdxsv

**Comment by vkedwardli** [comment]
Initial communication board emu is done on flycast: https://github.com/flyinghead/flycast/commit/a41a81f5dde460c711e4e2874bd20df16ae1ac01

But I think Derby Owners Club  & Gundam is not supported yet.

**Comment by p1pkin** [comment]
oh, nice, as I see it is based on my code from https://github.com/mamedev/mame/blob/master/src/mame/machine/m3comm.cpp
and of course does not preserve original license and copyright, smells like a typical flyinghead

@vkedwardli what you expecting from it ? it is actually my code, originate from Demul, which was kind of made public via MAME, and now was borrowed by flycast

**Comment by vkedwardli** [comment]
@p1pkin Thank you for your contribution, personally I'm happy that this open source work from you is picked up by others and it is beneficial to the gamers community. Not trying to start a emu comparison war, just posting what I know

**Comment by p1pkin** [comment]
@vkedwardli welcome. you probably don't know I'm a MAMEdev and Dumping Union member, so all the NAOMI/Atomiswave/etc games which was dumped and added to MAME (and later to reicast forks) during like last 5-7 years was made with my direct involvement and direction. not sure how many game sets it was, never counted them, probably several dozens. so, you can imagine how many personal time (and funds as well) was spent for the sake of ~~kids who want to play teh games~~ gamers comunity ;)

**Comment by vkedwardli** [comment]
I am not sure where the rage is from, your efforts are beneficial to the community, and have set the foundation for the platform. So please be happy if someone can build upon your success, I have not blamed you something like "Why You No release this feature" like an ETA son, I never intended to start a war seriously. If you don't release it, cool, if someone releases it, I'm happy, and I would love to share to the gamers above.

**Comment by p1pkin** [comment]
> I am not sure where the rage is from

just saying - there is no 'gamers community'. 'community' is the form of society where each person makes his own part of whole daily work. significant part of my life I was living in communistic/socialistic country, so I know 'a little' what I'm talking about - this gamers thing might be anything but community.
I'm personally see just a bunch of useless kids who want to play teh games for free ;)
